<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2020-10-27">

<title>Gram negative acquired resistance mechansims – Alessandro Gerada</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-c1fac2584b48ed01fb6e278e36375074.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed fullcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Alessandro Gerada</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../resources.html"> 
<span class="menu-text">Resources</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/agerada"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/alegerada"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Gram negative acquired resistance mechansims</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">microbiology</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">October 27, 2020</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p>The aim of this brief document is to summarise some general principles that allow the reader to identify acquired resistance mechanisms in Enterobacterales. For the purpose of this document this shall include organisms such as <em>E. coli</em>, <em>Klebsiella</em> spp., <em>Proteus</em> spp, <em>Enterobacter</em> spp, etc but importantly it will not extend to the nonfermenters (<em>Pseudomonas</em> spp., <em>Acinetobacter</em> spp., <em>Stenotrophomonas maltophilia</em>). This is not meant to be an exhaustive document - there exist other excellent papers that go into more detail, such as:</p>
<ul>
<li><a href="https://cmr.asm.org/content/22/1/161">CMR review on AmpC beta-lactamases</a></li>
<li><a href="https://cmr.asm.org/content/18/4/657#sec-53">CMR review on ESBLs</a></li>
<li><a href="https://www.eucast.org/fileadmin/src/media/PDFs/EUCAST_files/Resistance_mechanisms/EUCAST_detection_of_resistance_mechanisms_170711.pdf">EUCAST Resistance mechanisms</a></li>
</ul>
<p>Some notes on treatment options are also included, although again this is not meant to be an exhaustive list.</p>
</section>
<section id="extended-spectrum-beta-lactamases-esbl" class="level3">
<h3 class="anchored" data-anchor-id="extended-spectrum-beta-lactamases-esbl">Extended-spectrum beta-lactamases (ESBL)</h3>
<p>Knowing the definition of ESBL gets us halfway towards understanding how best to detect them. ESBLs are defined as:</p>
<blockquote class="blockquote">
<p>a group of β-lactamases which share the ability to hydrolyse third-generation cephalosporins and aztreonam yet are inhibited by clavulanic acid.</p>
</blockquote>
<p>Therefore this explains the use of Cefpodoxime as a screening test for ESBLs. The definitive phenotypic test is synergy testing with a beta-lactamase inhibitor. The principle behind these tests is important.</p>
<section id="screening-tests" class="level4">
<h4 class="anchored" data-anchor-id="screening-tests">Screening tests</h4>
<p>A third generation cephalosporin disc is the most sensitive and convenient. Cefpodoxime performs this role best.</p>
</section>
<section id="genotypic-tests" class="level4">
<h4 class="anchored" data-anchor-id="genotypic-tests">Genotypic tests</h4>
<p>There are some genotypic ESBL tests available. For example, some of the various <a href="https://www.biofiredx.com/products/the-filmarray-panels/filmarraybcid/">Biofire Filmarray</a> products can detect CTX-M. The main issue is that there are many different types and variations of ESBL (CTX, TEM, SHV, OXA-types), therefore the utility of a CTX-M PCR depends on whether this is the predominant gene in the tested population.</p>
</section>
<section id="phenotypic-tests" class="level4">
<h4 class="anchored" data-anchor-id="phenotypic-tests">Phenotypic tests</h4>
<p>These are probably the most useful confirmatory tests for ESBLs.</p>
<p>Enterobacterales that do <em>not</em> produce ESBL have a particular MIC or disc zone to third-generation cefalosporins. This is <em>not</em> affected by the presence of clavulanic acid.</p>
<p>Enterobacterales that <em>do</em> produce ESBL show (as per the original definition) a reduction of MIC (or increase in zone size) when the third-generation cefalosporin is combined with clavulanic acid.</p>
</section>
<section id="other-tests" class="level4">
<h4 class="anchored" data-anchor-id="other-tests">Other tests</h4>
<p>A chromogenic cefalosporin colorimetric test is available as a screening test, and follows the same principles as the chromogenic cefalosporin test for beta-lactamase detection in <em>Staph. aureus</em>.</p>
</section>
</section>
<section id="ampc-beta-lactamases" class="level3">
<h3 class="anchored" data-anchor-id="ampc-beta-lactamases">AmpC beta-lactamases</h3>
<p>The first point to make, and perhaps the most important, is that there are two types of AmpC beta-lactamases that we need to be concerned with:</p>
<ul>
<li>Chromosomal AmpC - these can be produced constitutively by certain bacterial species (such as <em>Enterobacter</em> spp., <em>C. freundii</em>, <em>M. morganii</em>, <em>P. aeruginosa</em>). These organisms in effect <strong>always</strong> have the ability to produce AmpC beta-lactamase through de-repression. Therefore whenever we isolate these organisms, we should assume that they can produce AmpC beta-lactamases, and should ignore any other results such as Cefoxitin susceptibility.</li>
</ul>
<blockquote class="blockquote">
<p>An interesting question - does <em>E. coli</em> have a chromosomal AmpC?</p>
</blockquote>
<ul>
<li>Acquired (usually plasmid mediated) AmpC - these have a similar effect to the above, but are acquired by bacterial species that otherwise do <em>not</em> have a chromosomal AmpC (organisms such as <em>Klebsiella</em> spp.). For these we do need further testing to determine their abscence or presence.</li>
</ul>
<p>All the points below only apply to acquired AmpC.</p>
<section id="screening-tests-1" class="level4">
<h4 class="anchored" data-anchor-id="screening-tests-1">Screening tests</h4>
<p>Cefoxitin is the most useful screening antibiotic for AmpC. However while it is very sensitive, it is not particularly specific, as there are other mechanisms of resistance against this antibiotic (such as porin loss). Therefore, we should also look at Cefotaxime and/or Ceftazidime. We should expect that if there is an acquired AmpC (rather than porin loss against Cefoxitin) that there is resistance to one of these agents as well.</p>
<p>In many laboratories, this is really where the AmpC story ends. Trying to confirm acquired AmpC further is high effort (especially as many isolates also co-produce ESBL) with low clinical yield, so many do not bother.</p>
<p>In summary (where “X” = any result):</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Organism</th>
<th>Cefoxitin</th>
<th>Cefotaxime/Ceftazidime</th>
<th>Interpretation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Has chromosomal AmpC</td>
<td>X</td>
<td>X</td>
<td>AmpC positive</td>
</tr>
<tr class="even">
<td>Does not have chromosomal AmpC</td>
<td>S</td>
<td>X</td>
<td>AmpC negative</td>
</tr>
<tr class="odd">
<td>Does not have chromosomal AmpC</td>
<td>R</td>
<td>S</td>
<td>AmpC negative</td>
</tr>
<tr class="even">
<td>Does not have chromosomal AmpC</td>
<td>R</td>
<td>R</td>
<td>AmpC positive</td>
</tr>
</tbody>
</table>
</section>
<section id="genotypic-tests-1" class="level4">
<h4 class="anchored" data-anchor-id="genotypic-tests-1">Genotypic tests</h4>
<p>PCR for different AmpC enzymes is available, but not in common routine use. Look elsewhere if you are curious.</p>
</section>
<section id="phenotypic-tests-1" class="level4">
<h4 class="anchored" data-anchor-id="phenotypic-tests-1">Phenotypic tests</h4>
<p>For exam purposes, it is worth mentioning that AmpC are inhibited by cloxacillin, giving similar results to the ESBL synergy tests if using cloxacillin instead of clavulanic acid, which is in theory a confirmatory test.</p>
</section>
</section>
<section id="carbapenemase-producing-enterobacterales-cpe" class="level3">
<h3 class="anchored" data-anchor-id="carbapenemase-producing-enterobacterales-cpe">Carbapenemase-producing Enterobacterales (CPE)</h3>
<p>This is a heterogenous group, and impossible to cover completely. Remember also that other mechanisms (such as AmpC de-repression with porin loss) are important causes of carbapenem resistance. It is however possible to learn a few tricks to distinguish between the 5 main classes:</p>
<ul>
<li>OXA-48</li>
<li>KPC</li>
<li>Metallo-beta-lactamases (MBL):
<ul>
<li>VIM</li>
<li>NDM</li>
<li>IMP</li>
</ul></li>
</ul>
<p>Fortunately if we group the metallo-beta-lactamases together we only have to deal with 3 groups (in the laboratory and even clinically VIM/NDM/IMP behave very similarly).</p>
<section id="screening-tests-2" class="level4">
<h4 class="anchored" data-anchor-id="screening-tests-2">Screening tests</h4>
<p>Ertapenem is the most sensitive antibiotic for CPE screening. Any zone smaller than 25mm in a suspect organism should be investigated further. Meropenem can also be used (better specificity at the cost of sensitivity).</p>
</section>
<section id="phenotypic-tests-2" class="level4">
<h4 class="anchored" data-anchor-id="phenotypic-tests-2">Phenotypic tests</h4>
<p>Until a few years ago, the Modified Hodge test was used as a phenotypic test for CPE. Thankfully (as this test is a bit tricky to explain) it is no longer recommended, due to poor performance.</p>
<p>Synergy tests can be used to identify CPE, and are quite popular exam questions:</p>
<ul>
<li>Synergy with boronic acid - <strong>KPC</strong> (although be aware that this also applies to AmpC)</li>
<li>Synergy with dipicolinic acid (DPA) or EDTA - <strong>MBL</strong></li>
<li>OXA-48 does not exhibit synergy with any agent</li>
</ul>
</section>
<section id="genotypic-tests-2" class="level4">
<h4 class="anchored" data-anchor-id="genotypic-tests-2">Genotypic tests</h4>
<p>These are by far the most practically useful tests to confirm the presence of CPE. <a href="https://www.cepheid.com/en/tests/Healthcare-Associated-Infections/Xpert-Carba-R">Cepheid Xpert Carba-R</a> detects the 5 main CPE classes quickly and with good results.</p>
</section>
<section id="other-tests-1" class="level4">
<h4 class="anchored" data-anchor-id="other-tests-1">Other tests</h4>
<p>It is possible to use colorimetric tests and even MALDI-TOF for the detection of CPE. However it is unlikely that these tests would appeal to a laboratory that has genotypic methods available.</p>
<p>Is it possible to summarise ESBL/AmpC/CPEs in one table? Yes, but at what cost?</p>
<table class="caption-top table">
<colgroup>
<col style="width: 22%">
<col style="width: 18%">
<col style="width: 36%">
<col style="width: 22%">
</colgroup>
<thead>
<tr class="header">
<th>Mechanism</th>
<th>Synergy</th>
<th>Antibiotic hints</th>
<th>Beta-lactam treatment options</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ESBL</td>
<td>Clavulanic acid ( + 3rd gen cefalosporin)</td>
<td>Cefpodoxime R <br> Cefepime R</td>
<td>Meropenem <br> Ertapenem <br> Temocillin <br> Ceftazidime-avibactam <br> Ceftolozane-tazobactam</td>
</tr>
<tr class="even">
<td>Acquired AmpC</td>
<td>Boronic acid or cloxacillin (+ 3rd gen cefalosporin)</td>
<td>Cefoxitin + Ceftazidime/Cefotaxime R <br> Cefepime S</td>
<td>Similar to ESBL</td>
</tr>
<tr class="odd">
<td>KPC</td>
<td>Boronic acid (+ carbapenem)</td>
<td></td>
<td>Meropenem-vaborbactam <br> Ceftazidime-avibactam <br> Imipenem-relebactam <br> Cefiderocol</td>
</tr>
<tr class="even">
<td>OXA-48</td>
<td>None</td>
<td>High level Temocillin R <br> Ceftazidime S</td>
<td>Ceftazidime-avibactam <br> Ceftazidime <br> Cefiderocol</td>
</tr>
<tr class="odd">
<td>MBL</td>
<td>DPA/EDTA (+ carbapenem)</td>
<td>Aztreonam S</td>
<td>Aztreonam-avibactam <br> Aztreonam <br> Cefiderocol</td>
</tr>
</tbody>
</table>
<p><strong>Important- this table does not apply if there are multiple resistance mechanisms.</strong></p>
<p>This table does not include non-beta-lactam classes of antibiotics, which can be used depending on susceptibility testing and source of infection. These agents, if sensitive, include:</p>
<ul>
<li>quinolones</li>
<li>aminoglycosides</li>
<li>polymixins</li>
<li>tigecycline</li>
<li>co-trimoxazole</li>
<li>chloramphenicol</li>
</ul>
</section>
</section>
<section id="conclusion-and-unanswered-questions" class="level3">
<h3 class="anchored" data-anchor-id="conclusion-and-unanswered-questions">Conclusion and unanswered questions</h3>
<p>There remain many unanswered questions in this field:</p>
<ul>
<li>Can we use piperacillin-tazobactam to treat ESBL/AmpC? <a href="https://pubmed.ncbi.nlm.nih.gov/30208454/">MERINO trial suggests not, at least for “definitive” therapy</a></li>
<li>Should we give combination therapy to all patients with invasive CPE infection, even if sensitive to one of the new beta-lactams? (probably, since resistance during treatment can occur with agents such as ceftazidime-avibactam)</li>
<li>Should we care about colonisation detected on IPC screens in patients who are unwell?</li>
</ul>
<p>It is also worth reviewing acquired colistin resistance (<em>mcr</em> genes) altough this seems to have gone quiet recently - possibly due to the introduction of new Gram-negative antibacterial agents having been released.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>